### PR TITLE
Introduced rx sc component and adapted endpoint explorer

### PR DESCRIPTION
**Pull request by distantcam**
I've created a reactive ServiceControl interface, which changes the interaction with SC from a pull to a push model. The main benefit here is a single call to SC which then broadcasts the data within SI rather than each viewmodel calling SC independently, which is causing SI to call SC multiple times.

The second part of this change is to rewrite the Endpoint Explorer to use this new interface. This lead to some code simplification.

### CODE REVIEW

File: src/ServiceInsight/Explorer/EndpointExplorer/EndpointExplorerViewModel.cs [hunk 1]
```diff
@@ -3,203 +3,97 @@
...
+        static EndpointExplorerViewModel()
         {
-            this.eventAggregator = eventAggregator;
-            this.settingsProvider = settingsProvider;
-            this.serviceControl = serviceControl;
-            this.networkOperations = networkOperations;
-            this.connectionProvider = connectionProvider;
-            this.commandLineParser = commandLineParser;
-            Items = new BindableCollection<ExplorerItem>();
+            serializer.ContractResolver = new SnakeCasePropertyNamesContractResolver();
```
Comment (HEskandari) [inline src/ServiceInsight/Explorer/EndpointExplorer/EndpointExplorerViewModel.cs]: I find the split a bit confusing...could this be merged with the initialiation code above? e.g.

    static EndpointExplorerViewModel()
    {
        serializer = new JsonSerializer { ContractResolver = new SnakeCasePropertyNamesContractResolver() }; 
    }

File: src/ServiceInsight/Explorer/EndpointExplorer/EndpointExplorerViewModel.cs [hunk 2]
```diff
@@ -3,203 +3,97 @@
...
-        }
+                Items.Add(root);
+            }
 
-        string GetConfiguredAddress()
-        {
-            if (commandLineParser.ParsedOptions.EndpointUri == null)
+            var toRemove = root.Children.ToList();
+            var endpointInstancesGroupedByName = e.Data.OrderBy(x => x.name).GroupBy(x => x.name);
+            foreach (var scaledOutEndpoint in endpointInstancesGroupedByName)
```
Comment (HEskandari) [inline src/ServiceInsight/Explorer/EndpointExplorer/EndpointExplorerViewModel.cs]: hmmm why is this called scaledOutEndpoint?

File: src/ServiceInsight/Explorer/EndpointExplorer/EndpointExplorerViewModel.cs [hunk 3]
```diff
@@ -3,203 +3,97 @@
...
-        void SelectDefaultEndpoint()
+        public void Handle(RequestSelectingEndpoint message)
         {
-            if (ServiceControlRoot == null)
+            foreach (var item in Items.OfType<ServiceControlExplorerItem>())
             {
-                return;
-            }
-
-            if (!commandLineParser.ParsedOptions.EndpointName.IsEmpty())
```
Comment (HEskandari) [inline src/ServiceInsight/Explorer/EndpointExplorer/EndpointExplorerViewModel.cs]: This code was required for when SI is activated via a link, e.g. si://ServiceControlUrl/endpoint and it'd focus on that particular endpoint. Was it removed intentionally?

File: src/ServiceInsight/Framework/RxServiceControl.cs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        void SetRefresh(TimeSpan interval);
+
+        void DisableRefresh();
+    }
+
+    class RxServiceControl : IRxServiceControl
+    {
+        static Serilog.ILogger anotarLogger = Log.ForContext<IRxServiceControl>();
+
+        public static IRxServiceControl Instance { get; set; }
```
Comment (HEskandari) [inline src/ServiceInsight/Framework/RxServiceControl.cs]: hmmmm isn't this registered as Singleton in the container? why do we need another singleton on top of that?

File: src/ServiceInsight/Startup/AppBootstrapper.cs [hunk 5]
```diff
@@ -27,6 +28,10 @@ public AppBootstrapper()
 
         protected override void Configure()
         {
+            BlobCache.ApplicationName = "ServiceInsight";
+
+            RxServiceControl.Instance = new RxServiceControl(BlobCache.UserAccount);
```
Comment (HEskandari) [inline src/ServiceInsight/Startup/AppBootstrapper.cs]: Same point as above...why reimplement singleton?

**Comment by HEskandari** [comment]
Made some comments. The only bigger reservation I have is how `JObject`s are passed around and scattered and the new `IRxServiceControl` although has fine grained methods (GetMessages, GetEndpoints) the results are not typed DTOs and are `dynamic` or `JObject`s.

File: src/ServiceInsight/Framework/RxServiceControl.cs (hunk 4 above)
Comment (distantcam) [inline src/ServiceInsight/Framework/RxServiceControl.cs]: No it's not registered as a singleton in the container.

I feel that with singletons we should start to take them out of the container to simplify the constructors for certain objects. I'm not sure about this. We should probably have a good discussion about it.

File: src/ServiceInsight/Explorer/EndpointExplorer/EndpointExplorerViewModel.cs (hunk 3 above)
Comment (distantcam) [inline src/ServiceInsight/Explorer/EndpointExplorer/EndpointExplorerViewModel.cs]: No this was unintentional. I forgot to check that this functionality remained. Let me check and see if I need to update this PR.

**Comment by distantcam** [comment]
> The only bigger reservation I have is how JObjects are passed around and scattered and the new IRxServiceControl although has fine grained methods (GetMessages, GetEndpoints) the results are not typed DTOs and are dynamic or JObjects.

Yep again, we should have a conversation about this.

My point of view is this, yes I used to be in the same boat as you, which is to convert things to a POCO. Over time however, those POCO's change. Common values are put into a base class, and helper methods are added, and suddenly we no longer have a POCO. Also why convert to a POCO? It's a layer between the data and the code that works on the data. What does the POCO give us for that layer?

File: src/ServiceInsight/Explorer/EndpointExplorer/EndpointExplorerViewModel.cs (hunk 3 above)
Comment (distantcam) [inline src/ServiceInsight/Explorer/EndpointExplorer/EndpointExplorerViewModel.cs]: Fixed this issue now

**Pull request merged**